File: assets/backing-services/java/tests/checkpoint_store_sqlite_test.java
package com.example.deliverystarter.adapters.driven.checkpointstoresqlite;

import com.example.deliverystarter.adapters.driven.eventstoresqlite.SqliteEventStore;
import com.example.deliverystarter.checkpointstorecontract.CheckpointStoreContract;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.AfterEach;

/**
 * The checkpoint contract, against SQLite — one file, one connection, real transactions.
 *
 * <p>This is the cheapest place the transactional checkpoint is genuinely proved: the in-memory adapter
 * rolls back by restoring a copy, whereas here a failed unit of work is a real ROLLBACK issued by a real
 * database. {@code :memory:} because the contract is about behaviour and a fresh database per case is what
 * keeps the cases independent.
 */
class SqliteCheckpointStoreTest extends CheckpointStoreContract {

    private final List<SqliteEventStore> opened = new ArrayList<>();

    @AfterEach
    void closeWhatWasOpened() {
        opened.forEach(SqliteEventStore::close);
        opened.clear();
    }

    @Override
    protected Stores stores() {
        SqliteEventStore store = new SqliteEventStore(":memory:");
        opened.add(store);
        return new Stores(store, new SqliteCheckpointStore(store));
    }
}

File: assets/backing-services/typescript/tests/checkpoint-store-sqlite.test.ts
import { createSqliteCheckpointStore } from '../../src/adapters/driven/checkpoint-store-sqlite.js';
import { openSqliteEventStore } from '../../src/adapters/driven/event-store-sqlite.js';
import { checkpointStoreContract } from './checkpoint-store-contract.js';

/**
 * The checkpoint contract, against SQLite — one file, one connection, real transactions.
 *
 * This is the cheapest place the transactional checkpoint is genuinely proved: the in-memory adapter rolls
 * back by restoring a copy, whereas here a failed unit of work is a real ROLLBACK issued by a real
 * database. `:memory:` because the contract is about behaviour and a fresh database per test is what keeps
 * its cases independent.
 */
checkpointStoreContract('the SQLite checkpoint store', async () => {
  const store = openSqliteEventStore(':memory:');
  return { store, checkpoints: createSqliteCheckpointStore(store) };
});
